# Assignment - Logistic Regression

In this assignment, you are going to use your model to predict survival of the Titanic disaster. For this assignment, send a link to a Jupyter notebook containing solutions to the following tasks.

- Download Titanic data from Kaggle. The data in the train.csv file meets your need.
- Split your data into training and test sets.
- Predict the survival based on the test data you split by creating your model.
- Is your model's performance satisfactory? Explain.
- Try to improve your model's performance by adding or subtracting some variables. 
- Explore the advantages and disadvantages of Logistic Regression and discuss with your mentor.

In [1]:
import pandas as pd
import numpy as np
train_data = pd.read_csv('titanic_train.csv')
test_data = pd.read_csv('titanic_test.csv')

In [2]:
train_data.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

In [3]:
train_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    object 
 11  Embarked     889 non-null    object 
dtypes: float64(2), int64(5), object(5)
memory usage: 83.7+ KB


In [4]:
test_data.head()

PassengerId  Pclass                                          Name     Sex  \
0          892       3                              Kelly, Mr. James    male   
1          893       3              Wilkes, Mrs. James (Ellen Needs)  female   
2          894       2                     Myles, Mr. Thomas Francis    male   
3          895       3                              Wirz, Mr. Albert    male   
4          896       3  Hirvonen, Mrs. Alexander (Helga E Lindqvist)  female   

    Age  SibSp  Parch   Ticket     Fare Cabin Embarked  
0  34.5      0      0   330911   7.8292   NaN        Q  
1  47.0      1      0   363272   7.0000   NaN        S  
2  62.0      0      0   240276   9.6875   NaN        Q  
3  27.0      0      0   315154   8.6625   NaN        S  
4  22.0      1      1  3101298  12.2875   NaN        S

In [5]:
train_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    object 
 11  Embarked     889 non-null    object 
dtypes: float64(2), int64(5), object(5)
memory usage: 83.7+ KB


In [6]:
for data in (train_data, test_data):
    is_male = pd.get_dummies(data['Sex'],drop_first = True)
    data.insert(loc=4, column='is_male', value=is_male)
    data.drop(labels='Sex', axis=1, inplace=True)    
del data

In [7]:
for data in (train_data, test_data):
    print(pd.unique(data[['Embarked']].values.ravel('K')))
    #print(pd.unique(data[['Cabin']].values.ravel('K')))

['S' 'C' 'Q' nan]
['Q' 'S' 'C']


In [8]:
# C = Cherbourg
# S = Southampton
# Q = Queenstown
for data in (train_data, test_data):
    data.Embarked = data.Embarked.replace('C', 1)
    data.Embarked = data.Embarked.replace('Q', 2)
    data.Embarked = data.Embarked.replace('S', 3)
    data.Embarked = data.Embarked.replace(np.nan, 0)
del data

In [9]:
for data in (train_data, test_data):
    data.dropna(subset=['Age'], how='all', inplace=True)
del data

In [10]:
train_data.info()
test_data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 714 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  714 non-null    int64  
 1   Survived     714 non-null    int64  
 2   Pclass       714 non-null    int64  
 3   Name         714 non-null    object 
 4   is_male      714 non-null    uint8  
 5   Age          714 non-null    float64
 6   SibSp        714 non-null    int64  
 7   Parch        714 non-null    int64  
 8   Ticket       714 non-null    object 
 9   Fare         714 non-null    float64
 10  Cabin        185 non-null    object 
 11  Embarked     714 non-null    float64
dtypes: float64(3), int64(5), object(3), uint8(1)
memory usage: 67.6+ KB
<class 'pandas.core.frame.DataFrame'>
Int64Index: 332 entries, 0 to 415
Data columns (total 11 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  332 non-null    int64  
 1 

In [11]:
print(train_data.columns, "\n", test_data.columns)

Index(['PassengerId', 'Survived', 'Pclass', 'Name', 'is_male', 'Age', 'SibSp',
       'Parch', 'Ticket', 'Fare', 'Cabin', 'Embarked'],
      dtype='object') 
 Index(['PassengerId', 'Pclass', 'Name', 'is_male', 'Age', 'SibSp', 'Parch',
       'Ticket', 'Fare', 'Cabin', 'Embarked'],
      dtype='object')


In [12]:
X = train_data[['Pclass','is_male','Age','SibSp','Parch','Fare','Embarked']]
y = train_data['Survived']

In [13]:
from sklearn.linear_model import LogisticRegression
log_reg = LogisticRegression()

## PS. Before, I have seen the queiston which says "The data in the train.csv file meets your need. Split your data into training and test sets.", I was going to use both train and test datasets on the Kaggle. I decided to keep EDA part on the test data and here is the train data allocation by using train_test_split

In [14]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test =  train_test_split(X, y, test_size=0.20, random_state=111)

In [15]:
log_reg.fit(X_train, y_train)

LogisticRegression()

# Model Performance Measurement
We can use .score () function to measure the performance of our model. This function returns the ratio of correct predictions. Let's check this for both training and test data.

In [16]:
train_accuracy = log_reg.score(X_train, y_train)
test_accuracy = log_reg.score(X_test, y_test)

print('One-vs.-Rest', '-'*30, 
      'Accuracy on Train Data : {:.2f}'.format(train_accuracy), 
      'Accuracy on Test Data  : {:.2f}'.format(test_accuracy), sep='\n')

One-vs.-Rest
------------------------------
Accuracy on Train Data : 0.80
Accuracy on Test Data  : 0.79


In [17]:
log_reg_mnm = LogisticRegression(multi_class='multinomial', solver='lbfgs')
log_reg_mnm.fit(X_train, y_train)

train_accuracy = log_reg_mnm.score(X_train, y_train)
test_accuracy = log_reg_mnm.score(X_test, y_test)

print('Multinomial (Softmax)', '-'*20, 
      'Accuracy on Train Data : {:.2f}'.format(train_accuracy), 
      'Accuracy on Test Data  : {:.2f}'.format(test_accuracy), sep='\n')

Multinomial (Softmax)
--------------------
Accuracy on Train Data : 0.79
Accuracy on Test Data  : 0.79


* When we check the results, we have 80% and 79% accuracy on train data. With test data, we have 79% and 79% accuracy. Both methods return approximately similar results.

In [18]:
import warnings; warnings.filterwarnings("ignore")

In [19]:
C_values = [0.001,0.01, 0.1,1,10,100, 1000]

accuracy_values = pd.DataFrame(columns=['C_values', 'Train Accuracy', 'Test Accuracy'])

for c in C_values:
    # Apply logistic regression model to training data
    lr = LogisticRegression(penalty = 'l2', C = c, random_state = 0, solver='lbfgs', multi_class='multinomial')
    lr.fit(X_train, y_train)
    accuracy_values = accuracy_values.append({'C_values': c,
                                              'Train Accuracy': lr.score(X_train, y_train),
                                              'Test Accuracy': lr.score(X_test, y_test)
                                             }, ignore_index=True)
display(accuracy_values)    

C_values  Train Accuracy  Test Accuracy
0     0.001        0.688266       0.692308
1     0.010        0.775832       0.748252
2     0.100        0.800350       0.790210
3     1.000        0.789842       0.790210
4    10.000        0.791594       0.790210
5   100.000        0.791594       0.790210
6  1000.000        0.791594       0.790210

- C value for 0.100 gives the best accuracy values!

# Second Model Performance Measurement

These two variables has been deleted.
- sibsp	# of siblings / spouses aboard the Titanic
- parch	# of parents / children aboard the Titanic

In [20]:
X2 = train_data[['Pclass','is_male','Age','Fare','Embarked']]
y2 = train_data['Survived']

In [21]:
X2_train, X2_test, y2_train, y2_test =  train_test_split(X2, y2, test_size=0.20, random_state=111)

In [22]:
log_reg.fit(X2_train, y2_train)

LogisticRegression()

In [23]:
train_accuracy2 = log_reg.score(X2_train, y2_train)
test_accuracy2 = log_reg.score(X2_test, y2_test)

print('One-vs.-Rest', '-'*30, 
      'Accuracy on Train Data : {:.2f}'.format(train_accuracy2), 
      'Accuracy on Test Data  : {:.2f}'.format(test_accuracy2), sep='\n')

One-vs.-Rest
------------------------------
Accuracy on Train Data : 0.80
Accuracy on Test Data  : 0.79


In [24]:
log_reg_mnm = LogisticRegression(multi_class='multinomial', solver='lbfgs')
log_reg_mnm.fit(X2_train, y_train)

train_accuracy2 = log_reg_mnm.score(X2_train, y_train)
test_accuracy2 = log_reg_mnm.score(X2_test, y_test)

print('Multinomial (Softmax)', '-'*20, 
      'Accuracy on Train Data : {:.2f}'.format(train_accuracy2), 
      'Accuracy on Test Data  : {:.2f}'.format(test_accuracy2), sep='\n')

Multinomial (Softmax)
--------------------
Accuracy on Train Data : 0.80
Accuracy on Test Data  : 0.79


* When we check the results, we have 80% and 80% accuracy on train data. With test data, we have 79% and 79% accuracy. Both methods return approximately similar results.

In [25]:
C_values = [0.001,0.01, 0.1,1,10,100, 1000]

accuracy_values2 = pd.DataFrame(columns=['C_values', 'Train Accuracy', 'Test Accuracy'])

for c in C_values:
    # Apply logistic regression model to training data
    lr = LogisticRegression(penalty = 'l2', C = c, random_state = 0, solver='lbfgs', multi_class='multinomial')
    lr.fit(X2_train, y2_train)
    accuracy_values2 = accuracy_values2.append({'C_values': c,
                                              'Train Accuracy': lr.score(X2_train, y2_train),
                                              'Test Accuracy': lr.score(X2_test, y2_test)
                                             }, ignore_index=True)
display(accuracy_values2)    

C_values  Train Accuracy  Test Accuracy
0     0.001        0.674256       0.671329
1     0.010        0.760070       0.727273
2     0.100        0.800350       0.811189
3     1.000        0.795096       0.790210
4    10.000        0.793345       0.783217
5   100.000        0.793345       0.783217
6  1000.000        0.793345       0.783217

- C value for 0.100 gives the best accuracy values!

# Second model gives the best results.
The EDA part on the first part has increased the performance of the modal.
I believe without the EDA part all results will be lower than now.
- Sex column is changed to is_male.
- Embarked is changed to integer.
- Rows whose Age variable is nan are dropped.